### PR TITLE
[SPIKE][Consensus] Remove validatorMap from the consensusModule struct - (Issue #331)

### LINKED ISSUE #331: [SPIKE][Consensus] Remove `validatorMap` from the `consensusModule` struct
## Objective

Remove `validatorMap` from the `consensusModule` struct state & access it from the persistence layer (added n the past) 

The `validatorMap` is just a placeholder for the global `ValidatorSet` which is stored at the persistence layer. 

*NOTE: this is a multi-module issue*

## Origin Document

Originally scoped in issue #315 - separate issue to downscope

## Goals
- [ ] Remove `validatorMap` from the `consensusModule` struct and port it to the persistence module's `GetActor(s)` (Validator) 
- [ ] Update Changelog.md

## Deliverable
- [ ] Code complete
- [ ] Supporting documentation

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

**Creator**: @andrewnguyen22

### PR DESCRIPTION
**Pull request by deblasis**
## Description

This PR builds upon #402 and removes `validatorMap` from the `consensusModule` struct, encapsulating the logic so that the validator set is dynamic and retrieved from the `persistence` module.

Also, since the change introduced the natural dependency on `persistence`, it made me question and revisit the way the modules register themselves with the `bus` and also how the modules are instantiated.

In an effort to simply the codebase and testing, while we improve the way we do dependency injection, the modules now are instantiated with a `bus` that will allow them to access to all the dependencies they need. 

Previously the modules required a `runtimeMgr` in their constructor and the bus registration was handled like an after-thought. 

## Issue

Fixes #331 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [x] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Removed `validatorMap` from `consensusModule` and encapsulated the logic so that it's dynamic, per-height and sourced from `persistence`
- Refactored module registration in `bus`
- Refactored modules instantiation and DI (`runtimeMgr` -> `bus`)
- The `runtimeMgr` is now a dependency and it can be accessed via the bus
- Minor improvements in tests

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by deblasis** [comment]
@Olshansk before I update CHANGELOGs and READMEs please take a look at this 🙏

File: telemetry/noop_module.go [hunk 1]
```diff
@@ -64,55 +63,55 @@ func (m *NoopTelemetryModule) GetEventMetricsAgent() modules.EventMetricsAgent {
 }
 
 func (m *NoopTelemetryModule) EmitEvent(namespace, event_name string, labels ...any) {
-	NOOP()
+	NOOP("EmitEvent", "namespace", namespace, "event_name", event_name, "labels", labels)
```
Comment (Olshansk) [inline telemetry/noop_module.go:66]: I think using reflections here is fair game to get the function name: https://stackoverflow.com/a/7053871/768439.

We can reuse it in all the functions

File: runtime/manager.go [hunk 2]
```diff
@@ -55,26 +61,36 @@ func NewManagerFromReaders(configReader, genesisReader io.Reader, options ...fun
...
 
 	return mgr
 }
 
-func NewManager(config *configs.Config, genesisState *genesis.GenesisState, options ...func(*Manager)) *Manager {
-	mgr := &Manager{
-		config:       config,
-		genesisState: genesisState,
-		clock:        clock.New(),
+func NewManager(config *configs.Config, genesis *genesis.GenesisState, options ...func(*Manager)) *Manager {
```
Comment (Olshansk) [inline runtime/manager.go:82]: Can we reuse `NewManager` in `NewManagerFromFiles`?

File: runtime/bus.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+func (m *bus) GetUtilityModule() modules.UtilityModule {
+	return m.modulesMap[modules.UtilityModuleName].(modules.UtilityModule)
+}
+
+func (m *bus) GetConsensusModule() modules.ConsensusModule {
+	return m.modulesMap[modules.ConsensusModuleName].(modules.ConsensusModule)
+}
+
+func (m *bus) GetTelemetryModule() modules.TelemetryModule {
+	telemetryModules := []string{modules.TelemetryModuleName, "telemetry_prometheus", "telemetry_noOP"}
```
Comment (Olshansk) [inline runtime/bus.go:80]: The postfix to `_prometheus` seems like something that should come from TelemetryConfigs, but it feels like scope creep in this PR. Can you add that as a TODO and just move `_prometheus` to a constant we can reuse in a couple of places?

**Review by Olshansk — CHANGES_REQUESTED** [review]
Don't know what to say other than the fact that this PR is 🔥

File: app/pocket/main.go [hunk 4]
```diff
@@ -22,8 +22,12 @@ func main() {
 	}
 
 	runtimeMgr := runtime.NewManagerFromFiles(*configFilename, *genesisFilename)
+	bus, err := runtime.CreateBus(runtimeMgr)
+	if err != nil {
+		log.Fatalf("Failed to create bus: %s", err)
+	}
 
-	pocketNode, err := shared.CreateNode(runtimeMgr)
+	pocketNode, err := shared.CreateNode(bus)
```
[the remark this answers is not in the document]
Comment (deblasis) [reply app/pocket/main.go:30]: Not necessary IMHO, it's going to be part of this PR :)

I didn't do it yet because I wanted a green light from you first.

File: consensus/helpers.go [hunk 5]
```diff
@@ -109,11 +121,15 @@ func isSignatureValid(msg *typesCons.HotstuffMessage, pubKeyString string, signa
...
+	validators, err := m.getValidatorsAtHeight(m.CurrentHeight())
+	if err != nil {
+		return err
+	}
+	return m.isOptimisticThresholdMet(len(m.messagePool[step]), validators)
 }
 
-func (m *consensusModule) isOptimisticThresholdMet(n int) error {
-	numValidators := len(m.validatorMap)
+func (m *consensusModule) isOptimisticThresholdMet(n int, validators []*coreTypes.Actor) error {
```
[the remark this answers is not in the document]
Comment (deblasis) [reply consensus/helpers.go:131]: I went with `numSignatures` passing the signatures directly would have resulted in unnecessary complexity in `hotstuff_replica.go`

File: consensus/helpers.go [hunk 6]
```diff
@@ -150,7 +166,15 @@ func (m *consensusModule) sendToNode(msg *typesCons.HotstuffMessage) {
...
 		return
 	}
-	if err := m.GetBus().GetP2PModule().Send(cryptoPocket.AddressFromString(m.idToValAddrMap[*m.leaderId]), anyConsensusMessage); err != nil {
+
+	validators, err := m.getValidatorsAtHeight(m.CurrentHeight())
+	if err != nil {
+		m.nodeLogError(typesCons.ErrPersistenceGetAllValidators.Error(), err)
+	}
+
+	_, idToValAddrMap := typesCons.GetValAddrToIdMap(validators)
```
[the remark this answers is not in the document]
Comment (deblasis) [reply consensus/helpers.go]: I see your point but I could argue that it wouldn't be DRY, the logic inside the two resulting methods would be basically the same. We would just return a map or the other.

image

If we had two methods like that you would have probably asked me to consolidate them :)

Maybe generics... but I feel like some reflection might be needed because `T ValAddrToIdMap | IdToValAddrMap` is not a core type (necessary to make the map).
It would probably look too complex for what it is simply a mapping....

I was about to reply with the above but I came up with this instead: https://github.com/pokt-network/pocket/pull/425/commits/fd6455dab9897bea61e62f00fd5cd5b435ce4838

image

File: consensus/helpers.go [hunk 7]
```diff
@@ -56,7 +57,18 @@ func (m *consensusModule) getQuorumCertificate(height uint64, step typesCons.Hot
...
 	}
 
-	if err := m.isOptimisticThresholdMet(len(pss)); err != nil {
+	persistenceReadContext, err := m.GetBus().GetPersistenceModule().NewReadContext(int64(height))
+	if err != nil {
+		return nil, err
+	}
+	defer persistenceReadContext.Close()
+
+	validators, err := persistenceReadContext.GetAllValidators(int64(height))
```
[the remark this answers is not in the document]
Comment (deblasis) [reply consensus/helpers.go:66]: Yep, sorry, I missed that one

File: consensus/hotstuff_leader.go [hunk 8]
```diff
@@ -303,17 +303,26 @@ func (m *consensusModule) validatePartialSignature(msg *typesCons.HotstuffMessag
...
 	address := partialSig.GetAddress()
-	validator, ok := m.validatorMap[address]
+
+	validators, err := m.getValidatorsAtHeight(m.CurrentHeight())
+	if err != nil {
+		return err
+	}
+
+	validatorMap := typesCons.ActorListToValidatorMap(validators)
+	valAddrToIdMap, _ := typesCons.GetValAddrToIdMap(validators)
```
[the remark this answers is not in the document]
Comment (deblasis) [reply consensus/hotstuff_leader.go]: Lemme know if this is acceptable instead https://github.com/pokt-network/pocket/pull/425#discussion_r1063302672

File: runtime/manager.go (hunk 2 above)
Comment (deblasis) [reply runtime/manager.go]: Absolutely

File: telemetry/noop_module.go (hunk 1 above)
Comment (deblasis) [reply telemetry/noop_module.go:66]: I am not sure I like it:

image

It is "cool" that we can rely on build-time stuff like a func instead of strings but honestly I feel like it would just add needless complexity here.

It can be made prettier by creating a wrapper func that strips out the namespace and whatever was added after the function name "-fm" 🤔

Not sure if it's worth it but I am open to do it if you feel strongly about it.

File: runtime/bus.go (hunk 3 above)
Comment (deblasis) [reply runtime/bus.go]: Maybe addressed in https://github.com/pokt-network/pocket/pull/425/commits/488eae716d685139488fc33569dac5748935ded4 

I could have done something fancy like use an init function per implementation that appended its name to the slice in the `telemetry` package but I think it's better to KISS. When/if we'll have 10 different implementations of the same module that's gonna be another story.

File: p2p/utils_test.go [hunk 9]
```diff
@@ -97,10 +98,15 @@ func waitForNetworkSimulationCompletion(t *testing.T, p2pModules map[string]*p2p
...
-func createP2PModules(t *testing.T, runtimeConfigs []modules.RuntimeMgr) (p2pModules map[string]*p2pModule) {
-	p2pModules = make(map[string]*p2pModule, len(runtimeConfigs))
-	for i, runtimeConfig := range runtimeConfigs {
-		p2pMod, err := Create(runtimeConfig)
+func createP2PModules(t *testing.T, buses []modules.Bus) (p2pModules map[string]*p2pModule) {
+	p2pModules = make(map[string]*p2pModule, len(buses))
+	for i := range buses {
+		p2pMod, err := Create(buses[i])
+
+		// normally it's not necessary to set the bus explicitly since it's done automatically when the Create calls `bus.RegisterModule`
```
[the remark this answers is not in the document]
Comment (deblasis) [reply p2p/utils_test.go:106]: 🤦‍♂️of course. Tnx buddy

File: runtime/bus.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
+package runtime
+
+import (
```
[the remark this answers is not in the document]
Comment (deblasis) [reply runtime/bus.go:3]: Of course, I'll re-request review when it's all done

File: p2p/utils_test.go [hunk 11]
```diff
@@ -161,33 +183,32 @@ func createMockGenesisState(t *testing.T, valKeys []cryptoPocket.PrivateKey) *ge
...
 	busMock.EXPECT().GetPersistenceModule().Return(persistenceMock).AnyTimes()
+	busMock.EXPECT().GetConsensusModule().Return(consensusMock).AnyTimes()
 	busMock.EXPECT().GetTelemetryModule().Return(telemetryMock).AnyTimes()
-
-	return busMock
 }
 
 // Consensus mock - only needed for validatorMap access
-func prepareConsensusMock(t *testing.T, genesisState *genesis.GenesisState) *modulesMock.MockConsensusModule {
+func prepareConsensusMock(t *testing.T, bus modules.Bus, genesisState *genesis.GenesisState) *modulesMock.MockConsensusModule {
```
[the remark this answers is not in the document]
Comment (deblasis) [reply p2p/utils_test.go:195]: Fair point, leftovers from a previous implementation where I was using real buses instead of mocks to make sure it all worked hehe.

File: p2p/module_raintree_test.go [hunk 12]
```diff
@@ -213,28 +216,49 @@ func testRainTreeCalls(t *testing.T, origNode string, networkSimulationConfig Te
...
+		expectedCall := networkSimulationConfig[valId]
+		expectedReads := expectedCall.numNetworkReads + 1
+		expectedWrites := expectedCall.numNetworkWrites
+		log.Printf("[valId: %s] expected reads: %d\n", valId, expectedReads)
+		log.Printf("[valId: %s] expected writes: %d\n", valId, expectedWrites)
+
+		wg.Add(expectedReads)
+		connMocks[valId] = prepareConnMock(t, valId, &wg, expectedCall.numNetworkReads)
+
+		wg.Add(expectedWrites)
```
[the remark this answers is not in the document]
Comment (deblasis) [reply p2p/module_raintree_test.go:244]: Yes, it does work if I group the `wg.Add` calls.

I fixed a bug here. It's not documented in the CHANGELOG yet but I guess you can understand it from the lines above

File: p2p/module.go [hunk 13]
```diff
@@ -37,28 +33,28 @@ type p2pModule struct {
...
 
-func Create(runtimeMgr modules.RuntimeMgr) (modules.Module, error) {
-	return new(p2pModule).Create(runtimeMgr)
+func Create(bus modules.Bus) (modules.Module, error) {
+	return new(p2pModule).Create(bus)
 }
 
-// IMPROVE: need to define a better pattern for dependency injection. Currently we are probably limiting ourselves by having a common constructor `Create(runtimeMgr modules.RuntimeMgr) (modules.Module, error)` for all modules.
-func CreateWithProviders(runtimeMgr modules.RuntimeMgr, addrBookProvider providers.AddrBookProvider, currentHeightProvider providers.CurrentHeightProvider) (modules.Module, error) {
+// IMPROVE: need to define a better pattern for dependency injection. Currently we are probably limiting ourselves by having a common constructor `Create(bus modules.Bus) (modules.Module, error)` for all modules.
```
[the remark this answers is not in the document]
Comment (deblasis) [reply p2p/module.go]: 🫡 #429

**Comment by deblasis** [comment]
@Olshansk PTAL 🙏

**Comment by deblasis** [comment]
Note about https://github.com/pokt-network/pocket/pull/425/commits/65dce5233516e4422b2adb9e606b15c59f13cd7e

We talked about it and without this change, consensus is unstable in `LocalNet` because of the `service_url` / `generic_param` situation. Actually, let me know if this fix should be a separate PR (hotfix).

File: build/docs/CHANGELOG.md [hunk 14]
```diff
@@ -7,15 +7,21 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
 - Removed `BaseConfig` from `configs`
 - Centralized `PersistenceGenesisState` and `ConsensusGenesisState` into `GenesisState`
+- Removed `is_client_only` since it's set programmatically in the CLI
 
 ## [0.0.0.1] - 2022-12-29
 
-- Updated configs with the missing value `max_mempool_count`
-- Added `is_client_only` to `config1.json` so that Viper knows it can be overridden. Done in the Makefile in `make client_connect`. Setting this can be avoided if we merge the changes in https://github.com/pokt-network/pocket/compare/main...issue/cli-viper-environment-vars-fix
+- Updated all `config*.json` files with the missing `max_mempool_count` value
```
Comment (Olshansk) [inline build/docs/CHANGELOG.md:18]: We have two `# [0.0.0.1]` updates. Remove the one below.

File: shared/CHANGELOG.md [hunk 15]
```diff
@@ -7,7 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
-## [0.0.0.9] - 2023-01-04
+## [0.0.0.10] - 2023-01-09
+
+- Updated modules constructor to accept a `bus` and not a `runtimeMgr` anymore
+- Registering modules with the `bus` via `RegisterModule` method
+
+## [0.0.0.9] - 2023-01-09
```
Comment (Olshansk) [inline shared/CHANGELOG.md:15]:
```suggestion
## [0.0.0.9] - 2023-01-04
```

File: docs/development/README.md [hunk 16]
```diff
@@ -186,25 +186,43 @@ $ make client_start && make client_connect
...
+│   │       └── proto                 # Protobuf representing the core types used throughout the codebase
+│   ├── crypto
+│   ├── docs
+│   │   └── flows
+│   ├── messaging                     # Messaging structs and functions
+│   │   └── proto
+│   └── modules                       # Shared modules definitions (interfaces)
+├── telemetry                         # Implementation of the Telemetry module
+├── utility                           # Implementation of the Utility module
+└── Makefile                          # [to-be-deleted] The source of targets used to develop, build and test
```
Comment (Olshansk) [inline docs/development/README.md:225]:
```suggestion
└── Makefile                          # The source of targets used to develop, build and test
```

File: docs/development/README.md [hunk 17]
```diff
@@ -186,25 +186,43 @@ $ make client_start && make client_connect
...
-├── rpc              # Implementation of the RPC module
-├── runtime          # Implementation of the Runtime module
-├── shared           # [to-be-refactored] Shared types, modules and utils
-├── telemetry        # Implementation of the Telemetry module
-├── utility          # Implementation of the Utility module
-├── Makefile         # [to-be-deleted] The source of targets used to develop, build and test
+├── app                               # Entrypoint to running the Pocket node and clients
+│   ├── client                        # Entrypoint to running a local Pocket debug client
+│   └── pocket                        # Entrypoint to running a local Pocket node
+├── bin                               # [currently-unused] Destination for compiled pocket binaries
```
Comment (Olshansk) [inline docs/development/README.md:192]:
```suggestion
├── bin                               # Destination for compiled pocket binaries
```

File: docs/development/README.md [hunk 18]
```diff
@@ -186,25 +186,43 @@ $ make client_start && make client_connect
...
+├── persistence                       # Implementation of the Persistence module
+├── rpc                               # Implementation of the RPC module
+├── runtime                           # Implementation of the Runtime module
+│   ├── configs                       # Configuration struct definitions
+│   │   └── proto                     # Protobuf representing the specific configuration of the various modules
+│   ├── defaults                      # Default values for the configuration structs
+│   ├── genesis
+│   │   └── proto                     # Protobuf representing the genesis state of the Pocket blockchain
+│   └── test_artifacts                # Componentry used for generating test artifacts such as particular genesis states used in testing
+├── shared                            # [to-be-refactored] Shared types, modules and utils
```
Comment (Olshansk) [inline docs/development/README.md:210]:
```suggestion
├── shared                            # Shared types, modules and utils
```

File: runtime/bus.go [hunk 19]
```diff
@@ -61,30 +57,46 @@ func (m *bus) GetEventBus() modules.EventsChannel {
...
+		return mod.(modules.ConsensusModule)
+	}
+	log.Fatalf("%s", ErrModuleNotRegistered(modules.ConsensusModuleName))
+	return nil
 }
 
 func (m *bus) GetTelemetryModule() modules.TelemetryModule {
-	telemetryModules := []string{modules.TelemetryModuleName, "telemetry_prometheus", "telemetry_noOP"}
-	for _, moduleName := range telemetryModules {
+	for _, moduleName := range telemetry.ImplementationNames {
```
Comment (Olshansk) [inline runtime/bus.go:93]: Nice

File: consensus/types/actor_mapper.go [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+	valAddresses := make([]string, 0, len(validators))
+	for _, val := range validators {
+		addr := val.GetAddress()
+		valAddresses = append(valAddresses, addr)
+		am.validatorMap[addr] = val
+	}
+	sort.Strings(valAddresses)
+
+	for i, addr := range valAddresses {
+		nodeId := NodeId(i + 1)
```
Comment (Olshansk) [inline consensus/types/actor_mapper.go:31]:
```suggestion
		nodeId := NodeId(i + 1) // TODO(#434): Improve the use of NodeIDs
```

File: consensus/types/actor_mapper_test.go [hunk 21]
```diff
@@ -0,0 +1,190 @@
+package types
```
Comment (Olshansk) [inline consensus/types/actor_mapper_test.go:1]: This is awesome. Not only are the tests well defined and extensive, it also makes it very clear what all the mappers do.

**Comment by Olshansk** [comment]
> Note about [65dce52](https://github.com/pokt-network/pocket/commit/65dce5233516e4422b2adb9e606b15c59f13cd7e)
> 
> We talked about it and without this change, consensus is unstable in `LocalNet` because of the `service_url` / `generic_param` situation. Actually, let me know if this fix should be a separate PR (hotfix).

Can you leave a comment on discord so the team knows? I think this is really close to getting over the finish line.

**Comment by deblasis** [comment]
@Olshansk 🙏getting there. Thanks for the valuable inputs as always

**Review by Olshansk — APPROVED** [review]
One tiny comment, but otherwise, let's 🚢 it!
